File: Data Structures & Algorithms/balanced-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    constructor() {
        this._isBalanced = true
    }
    isBalanced(root) {
        this._findDiffHeight(root)
        return this._isBalanced
    }

    _findDiffHeight(root) {
        let leftHeight = 0, rightHeight = 0

        if (!root) return 0

        if (root.left) {
            leftHeight = 1 + this._findDiffHeight(root.left)
        }

        if (root.right) {
            rightHeight = 1 + this._findDiffHeight(root.right)
        }

        if (Math.abs(leftHeight - rightHeight) > 1) {
            this._isBalanced = false
        }

        return Math.max(leftHeight, rightHeight)
    }
}

File: Data Structures & Algorithms/convert-bst-to-greater-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        int sum = 0;
        _convertBST(root, sum);
        return root;
    }

    private int _convertBST(TreeNode root, int sum) {
        if (root == null)
            return sum;
        if (root.right != null) {
            sum = _convertBST(root.right, sum);
        }
        sum += root.val;
        root.val = sum;
        if (root.left != null) {
            sum = _convertBST(root.left, sum);
        }

        return sum;
    }
}